ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/PeerDiscoveryAgent.java

```diff
@@ -53,282 +51,201 @@
 
     validateConfiguration(config);
 
-    final List<DiscoveryPeer> bootstrapPeers =
+    this.peerRequirement = peerRequirement;
+    this.peerBlacklist = peerBlacklist;
+    this.nodeWhitelistController = nodeWhitelistController;
+    this.bootstrapPeers =
         config.getBootstrapPeers().stream().map(DiscoveryPeer::new).collect(Collectors.toList());
 
-    this.vertx = vertx;
     this.config = config;
     this.keyPair = keyPair;
     this.peerTable = new PeerTable(keyPair.getPublicKey().getEncodedBytes(), 16);
-    this.controller =
-        new PeerDiscoveryController(
-            vertx,
-            this,
-            peerTable,
-            bootstrapPeers,
-            PEER_REFRESH_INTERVAL_MS,
-            peerRequirement,
-            peerBlacklist,
-            nodeWhitelistController);
+
+    id = keyPair.getPublicKey().getEncodedBytes();
   }
 
-  public CompletableFuture<?> start(final int tcpPort) {
-    final CompletableFuture<?> completion = new CompletableFuture<>();
+  protected abstract TimerUtil createTimer();
+
+  protected abstract CompletableFuture<InetSocketAddress> listenForConnections();
+
+  protected abstract CompletableFuture<Void> sendOutgoingPacket(
+      final DiscoveryPeer peer, final Packet packet);
+
+  public abstract CompletableFuture<?> stop();
+
+  public CompletableFuture<?> start() {
+    final CompletableFuture<?> future = new CompletableFuture<>();
     if (config.isActive()) {
       final String host = config.getBindHost();
       final int port = config.getBindPort();
       LOG.info("Starting peer discovery agent on host={}, port={}", host, port);
 
-      vertx
-          .createDatagramSocket(
-              new DatagramSocketOptions().setIpV6(NetworkUtility.isIPv6Available()))
-          .listen(
-              port,
-              host,
-              res -> {
-                if (res.failed()) {
-                  Throwable cause = res.cause();
-                  LOG.error("An exception occurred when starting the peer discovery agent", cause);
-
-                  if (cause instanceof BindException || cause instanceof SocketException) {
-                    cause =
-                        new PeerDiscoveryServiceException(
-                            String.format(
-                                "Failed to bind Ethereum UDP discovery listener to %s:%d: %s",
-                                host, port, cause.getMessage()));
-                  }
-                  completion.completeExceptionally(cause);
-                  return;
+      listenForConnections()
+          .thenAccept(
+              (InetSocketAddress localAddress) -> {
+                // Once listener is set up, finish initializing
+                this.localAddress = localAddress;
+                advertisedPeer =
+                    new DiscoveryPeer(
+                        id,
+                        config.getAdvertisedHost(),
+                        localAddress.getPort(),
```

Comment:
why do we have two of `localAddress.getPort()` here? is it related to:

```
final String host = config.getBindHost();
final int port = config.getBindPort();
```

Reply:
That's how it behaved before.  But it does seem wrong - I'll try to follow-up with a fix so I don't keep pushing stuff into this PR.